### PR TITLE
Fix infinite deploy loop on small stack edge case

### PR DESCRIPTION
**Pull request by BourgoisMickael**
#### Which issue(s) does this change fix?
#4004 

#### Why is this change necessary?
In case you have lots of parallel deploys that might throttle cloudformation (or network condition) and a small stack/changeset that could complete deployment before the CLI reach the `describe_stack_events` loop. (so a delay in the `get_last_event_time`)

#### How does it address the issue?
Rely on `execute_change_set` / `create_stack` / `update_stack` request date instead of last event date to start the describe loop. 
There should be no possible delay between the stack operation and the first `time_stamp_marker` for `descibe_stack_events` .
So no way to skip events anymore.

#### What side effects does this change have?
`get_last_event_time` is not used aymore
https://github.com/aws/aws-sam-cli/blob/6110cb948f3291ed1ffe064c3b7b8da1c8b9ceee/samcli/lib/deploy/deployer.py#L346
(should I delete the function ?)

#### Mandatory Checklist

- [x] Write/update unit tests
- [x] `make pr` passes

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0).

### CODE REVIEW

**Review by hawflau — APPROVED** [review]
LGTM!

File: samcli/lib/deploy/deployer.py [hunk 1]
```diff
@@ -444,7 +445,9 @@ def _is_root_stack_event(event: Dict) -> bool:
     def _check_stack_not_in_progress(status: str) -> bool:
         return "IN_PROGRESS" not in status
 
-    def wait_for_execute(self, stack_name: str, stack_operation: str, disable_rollback: bool) -> None:
+    def wait_for_execute(
+        self, stack_name: str, stack_operation: str, disable_rollback: bool, execution_time: float
```
Comment (hawflau) [inline samcli/lib/deploy/deployer.py:449]: Should we default `execution_time` to `time.time()? This make the method a bit more intuitive to use - If I don't need to check against a specific timestamp, I just don't provide this argument.

```suggestion
        self, stack_name: str, stack_operation: str, disable_rollback: bool, execution_time: float = time.time()
```

File: samcli/lib/deploy/deployer.py [hunk 2]
```diff
@@ -457,14 +460,16 @@ def wait_for_execute(self, stack_name: str, stack_operation: str, disable_rollba
             The type of the stack operation, 'CREATE' or 'UPDATE'
         disable_rollback : bool
             Preserves the state of previously provisioned resources when an operation fails
+        execution_time : float
+            Changeset execution time (current time)
```
Comment (hawflau) [inline samcli/lib/deploy/deployer.py:464]: Can we update the description to explain clearer about what this argument is for?

**Review by hawflau — CHANGES_REQUESTED** [review]

File: samcli/lib/deploy/deployer.py (hunk 2 above)
Comment (BourgoisMickael) [inline samcli/lib/deploy/deployer.py:464]:
```suggestion
            Time of the last stack change execution request (like `execute_change_set`, `update_stack`, `create_stack`)
            Prevents missing events if the event streaming is delayed from the change request by any action in between
```

Comment (BourgoisMickael) [reply samcli/lib/deploy/deployer.py]: I've improved the documentation

**Review by hawflau — APPROVED** [review]
Thanks for your contribution! LGTM

**Review by mildaniel — APPROVED** [review]

**Pull request merged**